core/staking_verifier.go

```diff
@@ -519,33 +569,40 @@ func VerifyAndMigrateFromMsg(
 //
 // Note that this function never updates the stateDB, it only reads from stateDB.
 func VerifyAndCollectRewardsFromDelegation(
-	stateDB vm.StateDB, delegations []staking.DelegationIndex,
+	stateDB vm.StateDB,
+	delegations []staking.DelegationIndex,
+	msg *staking.CollectRewards,
+	epoch *big.Int,
+	chainConfig *params.ChainConfig,
 ) ([]*staking.ValidatorWrapper, *big.Int, error) {
 	if stateDB == nil {
 		return nil, nil, errStateDBIsMissing
 	}
 	updatedValidatorWrappers := []*staking.ValidatorWrapper{}
 	totalRewards := big.NewInt(0)
 	for i := range delegations {
-		delegation := &delegations[i]
+		delegationIndex := &delegations[i]
 		// request a copy, and since delegations will be changed (.Reward.Set), copy them too
-		wrapper, err := stateDB.ValidatorWrapper(delegation.ValidatorAddress, false, true)
+		wrapper, err := stateDB.ValidatorWrapper(delegationIndex.ValidatorAddress, false, true)
 		if err != nil {
 			return nil, nil, err
 		}
-		if uint64(len(wrapper.Delegations)) > delegation.Index {
-			delegation := &wrapper.Delegations[delegation.Index]
-			if delegation.Reward.Cmp(common.Big0) > 0 {
-				totalRewards.Add(totalRewards, delegation.Reward)
-				delegation.Reward.SetUint64(0)
-			}
-		} else {
-			utils.Logger().Warn().
-				Str("validator", delegation.ValidatorAddress.String()).
-				Uint64("delegation index", delegation.Index).
-				Int("delegations length", len(wrapper.Delegations)).
-				Msg("Delegation index out of bound")
-			return nil, nil, errors.New("Delegation index out of bound")
+		var delegation *staking.Delegation
+		if delegation, _, err = staking.FindDelegationInWrapper(
+			msg.DelegatorAddress,
+			wrapper,
+			delegationIndex,
+			chainConfig,
+			epoch,
+		); err != nil {
+			return nil, nil, err
+		}
+		if delegation == nil {
+			continue
+		}
+		if delegation.Reward.Cmp(common.Big0) > 0 {
```

Comment:
Just realized that we should also check whether the delegation reward is "0" after CollectReward() right?